aspects/schema.go

```diff
@@ -0,0 +1,437 @@
+		}
+	}
+
+	// a map can have a "schema" constrain with specific values
+	if schema, ok := level["schema"]; ok {
+		var nextLevel map[string]json.RawMessage
+		if err := json.Unmarshal(schema, &nextLevel); err != nil {
+			return fmt.Errorf(`cannot unmarshal map's "schema" field: %v`, err)
+		}
+
+		v.entryTypes = make(map[string]Schema, len(nextLevel))
+		for key, value := range nextLevel {
+			validator, err := v.baseSchema.Parse(value)
+			if err != nil {
+				return fmt.Errorf(`cannot parse constraint for key %q: %w`, key, err)
+			}
+
+			v.entryTypes[key] = validator
+		}
+
+		return nil
+	}
+
+	// alternatively, it can constrain the type of its keys and/or values
+	keyDescription, ok := level["keys"]
+	if ok {
+		var err error
+		v.keyType, err = v.baseSchema.Parse(keyDescription)
+		if err != nil {
+			return fmt.Errorf(`cannot parse "keys" constraint in map schema: %w`, err)
+		}
+	}
+
+	valuesDescriptor, ok := level["values"]
+	if ok {
+		var err error
+		v.valueType, err = v.baseSchema.Parse(valuesDescriptor)
+		if err != nil {
+			return fmt.Errorf(`cannot parse "values" constraint in map schema: %w`, err)
+		}
+	}
+
+	return nil
+}
+
+type intSchema struct {
+	min     *int
+	max     *int
+	choices []int
+}
+
+func (v *intSchema) Validate(raw []byte) error {
+	var num int
+	if err := json.Unmarshal(raw, &num); err != nil {
+		return err
+	}
+
+	if v.min != nil && num < *v.min {
+		return fmt.Errorf(`integer %d is less than allowed minimum %d`, num, *v.min)
+	}
+
+	if v.max != nil && num > *v.max {
+		return fmt.Errorf(`integer %d is greater than allowed maximum %d`, num, *v.max)
+	}
+
+	if len(v.choices) != 0 {
+		var found bool
+		for _, choice := range v.choices {
+			if num == choice {
+				found = true
+				break
+			}
+		}
+
+		if !found {
+			return fmt.Errorf(`integer %d is not one of the allowed choices`, num)
+		}
+	}
+
+	return nil
+}
+
+func (v *intSchema) Parse(raw json.RawMessage) error {
+	constraints, err := getTypeConstraints(raw, "int")
+	if err != nil {
+		return err
+	} else if constraints == nil {
+		return nil
+	}
+
+	if rawChoices, ok := constraints["choices"]; ok {
+		var choices []int
+		err := json.Unmarshal(rawChoices, &choices)
+		if err != nil {
+			return err
+		}
+
+		if len(choices) == 0 {
+			return fmt.Errorf(`cannot have "choices" constraint with empty list: field must be populated or not exist`)
+		}
+
+		v.choices = choices
+	}
+
+	if minRaw, ok := constraints["min"]; ok {
+		var min int
+		if err := json.Unmarshal(minRaw, &min); err != nil {
+			return err
+		}
+		v.min = &min
+	}
+
+	if maxRaw, ok := constraints["max"]; ok {
+		var max int
+		if err := json.Unmarshal(maxRaw, &max); err != nil {
+			return err
+		}
+		v.max = &max
+	}
+
+	return nil
+}
+
+func getTypeConstraints(raw json.RawMessage, expectedType string) (map[string]json.RawMessage, error) {
+	var constraints map[string]json.RawMessage
+	if err := json.Unmarshal(raw, &constraints); err != nil {
```

Comment:
or maybe this JSON unarmashiling could be done one level up in CustomSchema.Parse ?